node/collation-generation/src/lib.rs

```diff
@@ -422,14 +375,277 @@ async fn handle_new_activations<Context>(
 	Ok(())
 }
 
-async fn obtain_current_validation_code_hash(
+#[overseer::contextbounds(CollationGeneration, prefix = self::overseer)]
+async fn handle_submit_collation<Context>(
+	params: SubmitCollationParams,
+	config: &CollationGenerationConfig,
+	ctx: &mut Context,
+	metrics: &Metrics,
+) -> crate::error::Result<()> {
+	let _timer = metrics.time_submit_collation();
+
+	let SubmitCollationParams {
+		relay_parent,
+		collation,
+		parent_head,
+		validation_code_hash_hint: code_hint,
+		result_sender,
+	} = params;
+
+	let validators = request_validators(relay_parent, ctx.sender()).await.await??;
+	let n_validators = validators.len();
+
+	// We need to swap the parent-head data, but all other fields here will be correct.
+	let mut validation_data = match request_persisted_validation_data(
+		relay_parent,
+		config.para_id,
+		OccupiedCoreAssumption::TimedOut,
+		ctx.sender(),
+	)
+	.await
```

Comment:
Why do runtime API requests return a `Receiver<Result<>>` rather than directly returning a `Result<>`?

Reply:
Under the hood it's implemented like

```rust
let (tx, rx) = oneshot::channel();
sender.send(RuntimeApiMessage::Request(at, RuntimeApiRequest::Foo(tx)).await;
rx
```